src/strategies/polygon/strategy-sushi-farm-base.sol

```diff
@@ -0,0 +1,158 @@
+        address _governance,
+        address _strategist,
+        address _controller,
+        address _timelock
+    )
+        public
+        StrategyBase(
+            _lp,
+            _governance,
+            _strategist,
+            _controller,
+            _timelock
+        )
+    {
+        poolId = _poolId;
+        token0 = _token0;
+        token1 = _token1;
+        IERC20(sushi).safeApprove(sushiRouter, uint(-1));
+        IERC20(weth).safeApprove(sushiRouter, uint(-1));
+        IERC20(wmatic).safeApprove(sushiRouter, uint(-1));
+    }
+    
+    function balanceOfPool() public override view returns (uint256) {
+        (uint256 amount, ) = IMiniChefV2(miniChef).userInfo(poolId, address(this));
+        return amount;
+    }
+
+    function getHarvestable() external view returns (uint256) {
+        return IMiniChefV2(miniChef).pendingSushi(poolId, address(this));
+    }
+
+    // **** Setters ****
+
+    function deposit() public override {
+        uint256 _want = IERC20(want).balanceOf(address(this));
+        if (_want > 0) {
+            IERC20(want).safeApprove(miniChef, 0);
+            IERC20(want).safeApprove(miniChef, _want);
+            IMiniChefV2(miniChef).deposit(poolId, _want, address(this));
+        }
+    }
+
+    function _withdrawSome(uint256 _amount)
+        internal
+        override
+        returns (uint256)
+    {
+        IMiniChefV2(miniChef).withdraw(poolId, _amount, address(this));
+        return _amount;
+    }
+
+    // **** Setters ****
+
+    function setKeepSUSHI(uint256 _keepSUSHI) external {
+        require(msg.sender == timelock, "!timelock");
+        keepSUSHI = _keepSUSHI;
+    }
+
+    // **** State Mutations ****
+
+    function harvest() public override onlyBenevolent {
+        // Anyone can harvest it at any given time.
+        // I understand the possibility of being frontrun
+        // But ETH is a dark forest, and I wanna see how this plays out
+        // i.e. will be be heavily frontrunned?
+        //      if so, a new strategy will be deployed.
+
+        // Collects SUSHI tokens
+        IMiniChefV2(miniChef).harvest(poolId, address(this));
+        uint256 _sushi = IERC20(sushi).balanceOf(address(this));
+        if (_sushi > 0) {
+            // 10% is locked up for future gov
+            uint256 _keepSUSHI = _sushi.mul(keepSUSHI).div(keepSUSHIMax);
+            IERC20(sushi).safeTransfer(
+                IController(controller).treasury(),
+                _keepSUSHI
+            );
+            _swapSushiswap(sushi, weth, _sushi.sub(_keepSUSHI));
+        }
+
+        // Collect MATIC tokens
+        uint256 _wmatic = IERC20(wmatic).balanceOf(address(this));
+        if (_wmatic > 0) {
+            _swapSushiswap(wmatic, weth, _wmatic);
+        }
+
+        // Swap half WETH for token0
+        uint256 _weth = IERC20(weth).balanceOf(address(this));
+        if (_weth > 0 && token0 != weth) {
+            _swapSushiswap(weth, token0, _weth.div(2));
```

Comment:
you did infinite approval of WETH in the constructor but set it to the small amount in line 129 as token0 is weth. So line 116 wouldn't work at all in the next `harvest`. So do `safeApprove` every time before swap rather than infinite approval.

Reply:
resolved